Allow provider sigalgs in SignatureAlgorithms conf

ajbozarth opened the pull request:
Though support for provider-based signature algorithms was added in ee58915 this functionality did not work with the SignatureAlgorithms configuration command. If SignatureAlgorithms is set then the provider sigalgs are not used and instead it used the default value.

This PR adds a check against the provider-base sigalg list when parsing the SignatureAlgorithms value.

Based-on-patch-by: Martin Schmatz <mrt@zurich.ibm.com>
Fixes #22761

ajbozarth commented:
I did not include any tests since, as far as I could find, there are no existing tests for this section of code. If I am wrong please point me at them and I will address it.

ajbozarth commented:
I just looked at the CLI failures and all three are the same failure. They seem to be related to my call of `SSL_CTX_free` but I am not yet able to figure out why the alert is failing at: https://github.com/openssl/openssl/blob/cf6342bc024868f5a55f2225f2e083415fb1329a/ssl/ssl_lib.c#L4132
The tests pass locally for me, and are only failing on those 3 CI

t8m commented:
**OTC: This came too late before 3.2.0. We will merge it later in 3.2.x.**

mattcaswell commented:
@martinschmatz - since you wrote the initial patch we will need a CLA from you and you will also need to be listed on the IBM CCLA.

See:

https://www.openssl.org/policies/cla.html

ajbozarth commented:
> OTC: This came too late before 3.2.0. We will merge it later in 3.2.x.

@t8m any rough idea on what the turnaround will be for 3.2.1? (ex Dec, Jan, Feb?)

I've also pinged @martinschmatz internally about the CLA issue

t8m commented:
> @t8m any rough idea on what the turnaround will be for 3.2.1? (ex Dec, Jan, Feb?)

Not later than after 3 months after the 3.2.0. However it can happen any time earlier if there are serious issues bug fixes or there are many of less serious ones.

ajbozarth commented:
I believe @martinschmatz has sent in his ICLA, as for his CCLA, if you need me to ping someone internally to get back to you I can. I believe @maximilien is our CCLA contact, but here's on vacation for the next week, so I would need to ping him to respond to any emails.

If there's anything else for me to do to "unblock" this just let me know

mattcaswell commented:
> I believe @maximilien is our CCLA contact, but here's on vacation for the next week, so I would need to ping him to respond to any emails.

Since this missed 3.2.0 anyway, I'd say let him have his vacation and sort it out when he's back!

> I did not include any tests since, as far as I could find, there are no existing tests for this section of code. If I am wrong please point me at them and I will address it.

There are extensive tests for various TLS related config options which are handled by test/ssl_test.c. All the configs that are tested are in the various test/ssl-tests/*.cnf files - which are themselves generated from cnf.in templates.

ssl_test is quite complicated. There is a whole README about it in the test sub-dir (see README.ssltest.md)

Unfortunately it doesn't look like we have any tests for SigAlgs set via config, so it seems to me we will need a new 33-sigalgs.cnf/22-sigalgs.cnf.in file

ssl-test also doesn't support a custom provider at the moment, so we'd need something for that - but presumably we can reuse the existing tls-provider that @baentsch mentioned which already has support for custom sigalgs.

It's quite likely you will also need to make changes to the test recipe that calls ssl_test in test/recipes/80-test_ssl_new.t

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2897,6 +2897,24 @@ static int sig_cb(const char *elem, int len, void *arg)
      */
     if (p == NULL) {
+        /* First check the providers */
+        SSL_CTX *ctx_helper = OPENSSL_zalloc(sizeof(*ctx_helper));
```
Style nit: Blank line after declarations

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2897,6 +2897,24 @@ static int sig_cb(const char *elem, int len, void *arg)
+        /* First check the providers */
+        SSL_CTX *ctx_helper = OPENSSL_zalloc(sizeof(*ctx_helper));
+        /* Load provider sigalgs */
+        if (!ctx_helper || !ssl_load_sigalgs(ctx_helper)) {
```
`ssl_load_sigalgs` reads `ctx->libctx` directly and `ctx->propq` indirectly (via `discover_provider_sig_algs`->`add_provider_sigalgs`) - neither of which have been set here. This is potentially quite problematic since `tls1_set_sigalgs_list` isn't passed this information either (although all callers of `tls1_set_sigalgs_list` do have it)

Probably a better solution is just to ensure the real SSL_CTX is passed down the call chain, rather than constructing a dummy one here.

ajbozarth replied:
I pushed an update that potentially addresses this and detailed it in a comment below. If you could give feedback on the "FEEDBACK QUESTION" I posed in that update.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2897,6 +2897,24 @@ static int sig_cb(const char *elem, int len, void *arg)
+        SSL_CTX *ctx_helper = OPENSSL_zalloc(sizeof(*ctx_helper));
+        /* Load provider sigalgs */
+        if (!ctx_helper || !ssl_load_sigalgs(ctx_helper)) {
+          ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
```
If `ssl_load_sigalgs` fails we don't free `ctx_helper` here

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2897,6 +2897,24 @@ static int sig_cb(const char *elem, int len, void *arg)
+              && strcmp(etmp, ctx_helper->sigalg_list[i].sigalg_name) == 0) {
+            sarg->sigalgs[sarg->sigalgcnt++] = ctx_helper->sigalg_list[i].code_point;
+            SSL_CTX_free(ctx_helper);
+            return 1;
```
returning 1 here means we are skipping the "reject duplicates" check below

mattcaswell reviewed (changes requested):
Don't we also need similar changes in `tls1_set_sigalgs`?

ajbozarth commented:
I'm back from vacation now and will start working on addressing @mattcaswell review and then into adding tests according to the guidance above.

ajbozarth commented:
I'm still working on this, it was sidelined last week, but my intention is to push the updates addressing review either Wed or Thur. As for the tests, though I intend to work on it more this week, I don't expect I will be able to finish that work before I go on my EOY holiday vacation. I'll make sure to put a final status update here at EOW, and then I'll return to work on the 2nd.

ajbozarth commented:
I pushed an update with the fixes that I've finished. I'm still working on passing the `ctx` from upstream, the upstream functions are called in quite a few places, so making sure changing the signature doesn't break anything is taking time.

ajbozarth commented:
So I'm still testing this, but I decided to push my update that passes the `SSL_CTX` down into the parser rather than creating a dummy (as suggested by @mattcaswell ).

This has caused the following changes to my PR:
- Due to the removal of the dummy I no longer have any alloc related calls
- I had to change the signature of a struct and a function to pass around the `SSL_CTX`
    - I added `SSL_CTX` to the `sig_cb_st` struct since that is just used for passing args to `sig_cb`
    - I added `SSL_CTX` as the first param to `tls1_set_sigalgs_list` since every other function that takes `SSL_CTX` as param always puts it first. I can change this
    - I kept the `SSL_CTX` and `CERT` params separate even though `SSL_CTX` contains the `CERT` value. 
        - ***FEEDBACK QUESTION***: The value of `CERT` is derived from two different source when calling `tls1_set_sigalgs_list`, either from `SSL_CTX` or from `SSL_CONNECTION`. *IF* `sc->cert` is the same as `sc->ssl->ctx->cert` then I could remove `CERT` as a param and just pass the `SSL_CTX` and use `ctx->cert` inside `tls1_set_sigalgs_list`
- I removed the `ERR_raise` since afaik it was only needed in case of a alloc failure (I will double check this with @martinschmatz )
- With the rearranging of the if statements the original check against the `sigalg_lookup_tbl` will run in case `SSL_CTX` is NULL (I threw an error before this change)

As I said I need to test this newest update more, as well as work on adding tests as asked. I will continue testing and looking at writing test Thur and Friday, but given how close the holiday break is I don't expect to push anymore code this month (unless I get feedback)

ajbozarth commented:
I was pulled onto another task yesterday that took up the remainder of my week, so my previous update is my final update for the year. I'll be back to this the first week of Jan, but I might check in on any responses over the break.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

ajbozarth commented:
> This PR is in a state where it requires action by openssl/otc but the last update was 30 days ago

I am still working on this (specifically learning the test framework so I can create new tests), but the majority of that 30 days I was on holiday

mattcaswell commented:
@ajbozarth - see #23391 which points out a memleak in this PR.

ajbozarth commented:
> @ajbozarth - see https://github.com/openssl/openssl/pull/23391 which points out a memleak in this PR.

I'll address this when I'm next in the code.

I've been trying the figure out the tests and how/where new ones for this file should be added, but my other work has been slowing down that progress. I am not sure how much longer it will take me to write the tests, and am open to collaboration on them if anyone desires to speed up merging this PR.

ajbozarth commented:
@mattcaswell I cherry-picked the commit from https://github.com/openssl/openssl/pull/23391 to give credit where due, but it seems that PR author isn't filed a CLA yet. I'm still working on tests and (with my conference travel schedule) I won't be done with those for weeks, so should I wait for them to submit a CLA or should I just force push an update adding the line of code under my own commit?

mattcaswell commented:
> @mattcaswell I cherry-picked the commit from https://github.com/openssl/openssl/pull/23391 to give credit where due, but it seems that PR author isn't filed a CLA yet. I'm still working on tests and (with my conference travel schedule) I won't be done with those for weeks, so should I wait for them to submit a CLA or should I just force push an update adding the line of code under my own commit?

The commit looks trivial to me. I would add "CLA: trivial" in a line on its own in the commit message for that commit.

ajbozarth commented:
> The commit looks trivial to me. I would add "CLA: trivial" in a line on its own in the commit message for that commit.

Thanks I've force pushed that update

t8m commented:
@ajbozarth CI is relevant

baentsch commented on ssl/t1_lib.c:
```diff
@@ -2897,11 +2899,27 @@ static int sig_cb(const char *elem, int len, void *arg)
+                              sarg->ctx->sigalg_list[i].sigalg_name) == 0) {
+                    sarg->sigalgs[sarg->sigalgcnt++] =
+                            sarg->ctx->sigalg_list[i].code_point;
+                    break;
```
What if `sarg->ctx->sigalg_list_len > OSSL_NELEM(sigalg_lookup_tbl)` and `i==OSSL_NELEM(sigalg_lookup_tbl)` on this `break`? I'd argue it'd be prudent to have tests for these (and other) cases.

ajbozarth replied:
Nice catch, I this stems from my reuse of the `i` var for both for loops. I'll add another incrementation var so each loop uses it's own

ajbozarth replied:
after a closer read of my code changes inside this block I found a few edge cases for how it could traverse the if blocks. So instead I've moved the `OSSL_NELEM` check inside the if block above it alongside the code previously in this section then separated the new if check into two steps to remove any unexpected edge cases due to fallen calls or for loops.

baentsch replied:
A test would arguably still be better than relying on visual inspection to find edge cases.

ajbozarth commented:
Update: I just pushed a fix addressing the most recent review. As for adding tests, my other work has made my time trying to understand the test suite slow. If my delay on adding tests becomes an issue feel free to ping me here and I'll work to reprioritize and speed up my development.

baentsch commented:
> Update: I just pushed a fix addressing the most recent review. As for adding tests, my other work has made my time trying to understand the test suite slow. If my delay on adding tests becomes an issue feel free to ping me here and I'll work to reprioritize and speed up my development.

I personally don't think it's an issue. It may be more efficient to set time aside to focus (e.g., getting this automatically/fully tested) rather than spending a little bit of time months apart to manually find corner cases, though.

beldmit commented:
@ajbozarth Would you mind rebasing this PR?

beldmit commented:
[27790_rebase.txt](https://github.com/openssl/openssl/files/14538395/27790_rebase.txt)
This is my attempt of rebasing, WFM

ajbozarth commented:
> Would you mind rebasing this PR?

Done, though I did not run any test or check yet, just did a merge with latest master and fixed the conflict. I'll double check it's still working when I return to this later.

t8m commented:
@ajbozarth Instead of merging the master branch, could you please do a rebase? `git reset --hard HEAD~ ; git rebase master ; git push --force`

We do not allow merge commits.

ajbozarth commented:
> We do not allow merge commits.

@t8m sorry, I checked the master git log and it seemed that all the PRs were being squashed prior to commit, so I though an easy merge was ok, I'll redo with a rebase when I next have a clean env (either tonight or tomorrow)

ajbozarth commented:
force pushed a rebased commit history instead of merged one as requested

ajbozarth commented:
@t8m I believe I've addressed your most recent review if you wanted to double check

romen reviewed (approved):
LGTM, great work!

beldmit commented:
Great work, many thanks!

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

beldmit commented:
Squashed and merged. Many thanks for the contribution!

ajbozarth commented:
Thanks for all the feedback and help on this

t8m commented:
Cherry-picked to 3.3 and 3.2 branches where it should have been applied as well.
